### ISSUE TITLE
Add support for SAP HANA's BOOLEAN data type

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#10152]
In the meantime, SAP HANA has added support for the `BOOLEAN` data type:
https://help.sap.com/viewer/4fe29514fd584807ac9f2a04f6754767/2.0.01/en-US/f044a9c7fb7b49159384f92f147345fb.html

We should support it natively, in jOOQ, as well.

**Comment by lukaseder** [comment]
The boolean type support is not complete. We can do this:

```sql
create table t (b boolean);
insert into t values (true);
insert into t values (false);
insert into t values (null);
select * from t where b = true;
```

But not this:

```sql
select * from t where b;
select 1 = 1 from t;
```

**Comment by niiick1** [comment]
Hi @lukaseder, is there a workaround for current version until this is implemented?

**Comment by lukaseder** [comment]
@niiick1 - I'm not sure if a generic workaround that works in all syntactic cases is possible, but some use-cases could be. First off, you can implement a custom data type binding: https://www.jooq.org/doc/latest/manual/sql-building/queryparts/custom-bindings, and attach that to all boolean types in your generated code. That binding can then take care of producing the desired `BOOLEAN` literal rather than `1/0`.

I'll be very happy to try to think of additional workarounds for specific use-cases that you may have.

**Comment by niiick1** [comment]
Yeah, I ended up using custom bindings, although it was a bit confusing when reading this document where it the custom binding would be used in the API, but it worked just fine.

Unfortunately, it's a little troublesome having to declare the binding on every query I need a boolean literal, as we are using jOOQ without code generation. I would need to warn the team when dealing with booleans, to use an abstraction instead of naturally using the jOOQ API.

I wonder if there is a way to globally replace the boolean binding for a SQL dialect, where using something like `.eq(true)` would use the custom binding instead of the default.

**Comment by lukaseder** [comment]
> Yeah, I ended up using custom bindings, although it was a bit confusing when reading this document where the custom binding would be used in the API, but it worked just fine.

Hm, I had linked the documentation about the binding, instead of the one about using it in code generation: https://www.jooq.org/doc/3.13/manual/code-generation/custom-data-type-bindings/

Does this help? This way, you can attach it automatically to every query without needing to worry about it.

**Comment by niiick1** [comment]
We are not using code generation. I assume there is no way to bind it as default then?

**Comment by lukaseder** [comment]
That would depend on how you create your queries. How do you do that?

**Comment by niiick1** [comment]
Generally, `DSL.using(configuration).select()`, where configuration sets the dialect and the render mode nothing else.

**Comment by lukaseder** [comment]
I meant how do you create table / field references? Do you create ad-hoc `DSL.field("xyz", Boolean.class)` references all the time, or do you have some sort of dynamically created, globally available meta data, where you could attach a data type `Binding` only once?

**Comment by niiick1** [comment]
Yeah, we use `DSL.field()`, no dynamic metadata available. We do use a JPA ORM, but I don't think we could use the same metadata, right?

**Comment by lukaseder** [comment]
> Yeah, we use `DSL.field()`, no dynamic metadata available

So, a workaround to support the `BOOLEAN` type on a per-`DSL.field()` usage would be to attach a custom data type bidning to the expressions *every time*. Something like this:

```java
// Can be reused, globally
DataType<Boolean> HANA_BOOLEAN = SQLDataType.BOOLEAN.asConvertedDataType(new MyHanaBooleanBinding());

// Attach the data type to each individual field:
Field<Boolean> field = field(name("name"), HANA_BOOLEAN);
```

> We do use a JPA ORM, but I don't think we could use the same metadata, right?

It's possible to generate jOOQ code from JPA meta data directly using the `JPADatabase`: https://www.jooq.org/doc/latest/manual/code-generation/codegen-jpa/

In any case, I strongly recommend using jOOQ with the code generator. There are many many advantages that will be available to you, including a simple workaround for this feature.

**Comment by lukaseder** [comment]
Implementing this now

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14. The implementation will be available with the next set of nightly builds from here:
https://www.jooq.org/download/versions

**Comment by niiick1** [comment]
Awesome, I think I'll go with the field approach for now. 

Thanks for the quick responses @lukaseder!
